### PR TITLE
[5.3] Ajax component support of Stringable results

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for https://github.com/joomla/joomla-cms/discussions/43471 .

### Summary of Changes

~~Adding support of `Stringable` https://www.php.net/manual/en/class.stringable.php results to Ajax component.~~
Adding `Joomla\CMS\String\StringableInterface` results to Ajax component, this is transitioning interface to  `Stringable` https://www.php.net/manual/en/class.stringable.php.
This allows to customise the response much better, than it is currently.

### Testing Instructions

Add following code in to `plugins/system/skipto/src/Extension/Skipto.php`:
```php
    public function onAjaxSkiptoTest1(\Joomla\CMS\Event\Plugin\AjaxEvent $event)
    {
        $event->updateEventResult(new class ('test 1') extends \Joomla\CMS\Response\JsonResponse implements \Joomla\CMS\String\StringableInterface {});
    }

    public function onAjaxSkiptoTest2(\Joomla\CMS\Event\Plugin\AjaxEvent $event)
    {
        $event->addResult(['foo', 'bar']);
    }

    public function onAjaxSkiptoTestError(\Joomla\CMS\Event\Plugin\AjaxEvent $event)
    {
        throw new \Exception('test error');
    }
```

Then open following links:
1) `/administrator/index.php?option=com_ajax&plugin=skiptoTest1&group=system&format=raw`
2) `/administrator/index.php?option=com_ajax&plugin=skiptoTest1&group=system&format=json`
3) `/administrator/index.php?option=com_ajax&plugin=skiptoTest2&group=system&format=json`
4) `/administrator/index.php?option=com_ajax&plugin=skiptoTestError&group=system&format=raw`
5) `/administrator/index.php?option=com_ajax&plugin=skiptoTestError&group=system&format=json`

### Expected result AFTER applying this Pull Request
Check response for each:
1) `{"success":true,"message":null,"messages":null,"data":"test 1"}`
2) `{"success":true,"message":null,"messages":null,"data":"test 1"}`
3) `{"success":true,"message":null,"messages":null,"data":[["foo","bar"]]}`
4) `Exception: test error`
5) `{"success":false,"message":"test error","messages":null,"data":null}`

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/274
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: components/com_ajax/ajax.php [hunk 1]
```diff
@@ -205,15 +205,19 @@
 // Return the results in the desired format
 switch ($format) {
     case 'json':
-        // JSONinzed
-        echo new JsonResponse($results, null, false, $input->get('ignoreMessages', true, 'bool'));
+        if (!($results instanceof Throwable) && $results instanceof Stringable) {
+            echo $results;
+        } else {
+            // JSONinzed
```
Comment (brianteeman) [inline components/com_ajax/ajax.php]:
```suggestion
            // JSONized
```

**Comment by korenevskiy** [comment]
Thanks, I tested it. I spent many hours testing on my project.
It can be seen that you spent just as many hours writing, testing, finding bugs and writing documentation for PR.
I tested in JSON mode, with error, without error, with data and with a Json Response object. I also set the parameters $input->set('ignoreMessages',true, 'bool'), and also tested in RAW mode, also in different configurations.
Now Joomla has become better. Now we can send customized messages as system messages in case of an error, and special notifications. Respect.

**Comment by Fedik** [comment]
@korenevskiy Please visit https://issues.joomla.org/tracker/joomla-cms/43530 and push the button, thanks
Screenshot 2024-05-29_23-32-36

**Comment by carlitorweb** [comment]
`/administrator/index.php?option=com_ajax&plugin=skiptoTest2&group=system&format=raw`

```log
Warning: Array to string conversion in [ROOT]\components\com_ajax\ajax.php on line 238
```

**Comment by carlitorweb** [comment]
I have tested this item :white_check_mark: successfully on 66f35b7956067b5a31978a21f4948b07a1ad28ee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Comment by korenevskiy** [comment]
I have tested this item :white_check_mark: successfully on 66f35b7956067b5a31978a21f4948b07a1ad28ee<br /><br />Thanks, I tested it. I spent many hours testing on my project.
It can be seen that you spent just as many hours writing, testing, finding bugs and writing documentation for PR.
I tested in JSON format, with error, without error, with data and with a JsonResponse object. I also set the parameters
$input->set('ignoreMessages',false, 'bool'), and also tested in RAW mode, also in different configurations.
I also made an error throw with a message.
Now Joomla has become better. Now we can send customized messages as system messages in case of an error, and special notifications. Respect.
### there were no problems during testing.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Comment by rdeutz** [comment]
We discussed this last week and we think this is a b/c break. It would be better to add a new format instead of changing an existing one and do "kind of magic" on this.

**Comment by Fedik** [comment]
> we think this is a b/c break

What kind of b.c break? I do not see any

**Comment by korenevskiy** [comment]
> We discussed this last week and we think this is a b/c break. It would be better to add a new format instead of changing an existing one and do "kind of magic" on this.

By creating a new format, we force each developer to refer to the documentation each time in order to understand the subtleties of the differences between each format. But ultimately the formats are the same, it's just a matter of processing JSON.

**Comment by Fedik** [comment]
p<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Comment by Fedik** [comment]
I have update PR, to be more backward compatible.
Please test again.

**Comment by korenevskiy** [comment]
Can you tell me how the `Joomla\CMS\String\StringableInterface` interface is used here?

It can be seen that the minimum version of Joomla 5.1 is 8.1, Joomla 5.2 may have the same.
The built-in `Stringable` interface in PHP has a minimum version of 8.0. 

Using the  `Joomla\CMS\String\StringableInterface`, you will oblige developers to repeatedly return to the documentation or to the AJAX component code to understand what needs to be inherited from the new interface.
If you inherit from  `Joomla\CMS\String\StringableInterface`, the extension will still depend on the latest version of Joomla 5.2.
And also the main reason for this whole idea.


The object of the `JsonResponse` class will be passed as a result to the parameter of the new object of the `JsonResponse` class and will be nested JSON in the main JSON. Thus, it does not give me the opportunity to debug the code by passing additional messages.


изображение

**Comment by korenevskiy** [comment]
Let's start with the fact
```php
if (!($results instanceof Throwable) && $results instanceof StringableInterface)
```
This
```php
if (!($results instanceof Throwable) && $results instanceof \Joomla\CMS\Response\JsonResponse)
```

Thus, we will allow those who want to additionally serialize a string in JSON to do so. And also independently transmit an object of class `\JsonResponse`.
or do a `json_validate()` check

**Comment by Fedik** [comment]
You can see in the test:
```php
new class ('test 1') extends \Joomla\CMS\Response\JsonResponse implements \Joomla\CMS\String\StringableInterface {};
```

>  the extension will still depend on the latest version of Joomla 5.2

yeap, it is new feature

**Comment by korenevskiy** [comment]
Why don't you just allow the `\JsonResponse` object to be returned from the user's extension so that it has `JSON`?

**Comment by korenevskiy** [comment]
I understood your idea, and I see the potential, but there is no simplicity in this solution. And if you allow the `\JsonResponse` object to be returned, the result will be the same.
Maybe?
```php
if (!($results instanceof Throwable) && ($results instanceof \Joomla\CMS\Response\JsonResponse || \Joomla\CMS\String\StringableInterface))
```

**Comment by korenevskiy** [comment]
Still, the new interface needs to be renamed so that it is clear that this applies to `\JsonResponse`, since the wrong interface will be inherited by mistake. And if suddenly the new year is intentionally inherited from the new interface in cases not related to `\JsonResponse`, then this will produce meaningless objects, as a result it will need to be inherited from 2 interfaces.
You have created a good flexibility, deeper, but it will cause problems. Either allow `\JsonResponse` to be returned, or return an object of a special interface from which `\JsonResponse` will be inherited. i.e. inherit `\JsonResponse` also from the `\Joomla\CMS\String\StringableInterface` interface

**Comment by korenevskiy** [comment]
> The idea is the response can be anything, as long as it is Stringable, and so can be rendered in browser.
> Example for `?format=xml`:

Of course, the string can be XML. But we are discussing the return inside the JSON condition. If the condition is JSON, then we need to check for JSON, or for JsonRespone

**Comment by korenevskiy** [comment]
I just realized that there is no point in introducing a new interface. Before, it seemed to me that `$var instanceof Stringtable` is equivalent to `is_string($var)'. I thought that supposedly both variables to cast a string would give the same value. But it's obvious that the values are the same, but the condition is different.
This means that there is no point in introducing a new interface.
The condition is if the extension code returns an object `$obj instanceof Stringable`, which means the object will be cast to a string, but if a string is returned from the extension code, then obviously it needs to be passed as a parameter to the object of the `JsonResponse` class.
There is no point for a new interface.

**Comment by Fedik** [comment]
>  But it's obvious that the values are the same

No. String is a string and Object is object (even when it implement Stringable).
`new JsonResponse($string)` will produce: `{"data":"the string value"}`
`new JsonResponse($stringableObject)` will produce : `{"data":{ ... the object properties}}`

**Comment by korenevskiy** [comment]
The answer can be anything, including XML. I agree, maybe, but it will be another condition of the format, are we discussing `?format=json`, and json creation takes place inside the switch case 'Json' condition.
Inside the JSON condition, the `\JSONResponse` object was created before, which means backward compatibility is quite simple, which is placed in the `\JSONResponse` object, the parameter should not be a `\JSONResponse`, this will already indicate that backward compatibility has been implemented.
Suppose you want new features to use any object to generate JSON, such an object can inherit the Stringable interface. In any case, there was no such opportunity before. The custom object will be converted to a string instead of a non-`\JSONResponse`. But if the user wants to bring his object with the interface to a string and additionally serialize, then let his object itself lead to a string, and already return the string from the extension, this will allow the string to be serialized inside `\JSONResponse`.

**Comment by Fedik** [comment]
>  which means backward compatibility is quite simple

Nope, when existing extension provides result as `JsonResponse`, then it will be wrapped in to another `JsonResponse`. So in result we will have:
`new JsonResponse($anotherJsonResponse)` will produce : `{"data":{"data":{ ... the object properties}}}`.

Even if it does not make sense, it still changing the response that com_ajax will send. and that will break that extension.

**Comment by korenevskiy** [comment]
> `new JsonResponse($anotherJsonResponse)` will produce : `{"data":{"data":{ ... the object properties}}}`.

```php
if (!($results instanceof Throwable) && $results instanceof Stringable)
```
At the same time, it is necessary to add to the `\JSONResponse`e class
`libraries/src/Response/JsonResponse.php`
```php
class JsonResponse implements  Stringable{
}
```



If you do as you say, parents should do this interface for `\JSONResponse`.

**Comment by korenevskiy** [comment]
I like your idea, but it seems to me that it is possible to achieve an elegant solution without adding a new interface.

**Comment by korenevskiy** [comment]
OR
```php
if (!($results instanceof Throwable) && $results instanceof StringableResponseInterface)
```
At the same time, it is necessary to add to the `\JSONResponse`e class
`libraries/src/Response/JsonResponse.php`
```php
class JsonResponse implements  StringableResponseInterface{
}
```

**Comment by Fedik** [comment]
No, `JsonResponse` already implements `Stringable` **implicitly**.
> Unlike most interfaces, Stringable is implicitly present on any class that has the magic __toString() method defined

And currently we cannot add our `StringableInterface` to it, because it will change response format. In reasons I wrote earlier.

Currently developers who wants a custom response should implement `StringableInterface` on their own. And in future (~joomla 7), we change com_ajax to use PHP native `Stringable`, and all will continue to work without breaks.

**Comment by korenevskiy** [comment]
Is it worth checking the object for   `JsonSerializable` interface?

**Comment by Fedik** [comment]
That a bit diferent approach, but could work in some cases also.

**Comment by korenevskiy** [comment]
Why is this method bad?
https://github.com/joomla/joomla-cms/pull/43530#issuecomment-2178436180

**Comment by Fedik** [comment]
Because as for now it behaviors as "double wrapping", changing it will be b/c break.
https://github.com/joomla/joomla-cms/pull/43530#issuecomment-2178233570

**Comment by korenevskiy** [comment]
```php
if (!($results instanceof Throwable) 
&& ($results instanceof \Joomla\CMS\Response\JsonResponse || \StringableResponseInterface)){
echo $results;
}
else{
echo new JsonResponse($results, null, false, $input->get('ignoreMessages', true, 'bool'));
}
```
Where `StringableResponseInterface` it is joomla new interface

### But where is the double explosion here?, because if the object is JsonResponse, then there is no "double wrapping". In this case, the conversion to the string type occurs.

**Comment by Fedik** [comment]
> because if the object is JsonResponse, then there is no "double wrapping"

will be, but **current behavior** (in j3, j4, j5) is diferent

**Comment by korenevskiy** [comment]
The previous behavior caused a nested object of class `\JsonResponse` to be created in a new amendment object of class `\JsonResponse`, now this object will be cast to class `\JsonResponse` .
Do you think there are people who returned the `\JsonResponse` object from the extension, which was re-serialized in the `\JsonResponse` object? If there are any, it will cause an error.
Well, then there is no solution.
Because the right decision will cause an error or misunderstanding.
Then let's assign this amendment to` Joomla 6`. Where it will be explicitly stated that nested `\JsonResponse` should not be in each other.

**Comment by Fedik** [comment]
That what this PR is doing. 
You already can use it with https://github.com/joomla/joomla-cms/pull/43530#issuecomment-2176411656 (if the PR will be merged). Then your code continue to work in j7, without changes.

**Comment by korenevskiy** [comment]
I know for sure that the interface name should indicate that the new interface was created to respond to the request.
There will be confusion in assigning the interface name, people will mistakenly inherit from the wrong interface, or for backward compatibility, you will have to store this file for a long time.
`StringableResponseInterface`

**Comment by Fedik** [comment]
The full name `Joomla\CMS\String\StringableInterface`, so it is fine

**Comment by korenevskiy** [comment]
What value will this interface have other than the Ajax class of the component?

**Comment by exlemor** [comment]
I have tested this item :white_check_mark: successfully on 4ea13497be08676e78c6a7a5a089ed636d437617<br /><br />I have tested this successfully. ( Thanks to @LadySolveig for making sure I didn't kill another test server ;) lol )<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.3-dev.

**Comment by fgsw** [comment]
I have tested this item :white_check_mark: successfully on 4ea13497be08676e78c6a7a5a089ed636d437617<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Comment by alikon** [comment]
rtc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43530">issues.joomla.org/tracker/joomla-cms/43530</a>.</sub>

**Pull request merged**